types/sinon/ts3.1/index.d.ts

```diff
@@ -1155,129 +1155,129 @@ declare namespace Sinon {
          * Passes if spy was never called
          * @param spy
          */
-        notCalled(spy: SinonSpy): void;
+        notCalled(spy: SinonSpy<any>): void;
```

Comment:
what happened here? `SinonSpy` defaults its generics to `any`

Reply:
No, it defaults the parameter types to `any[]`. `any[]` is not assignable to tuples, because `any[]` is arbitrary length (0-n) while a tuple requires a fixed amount of members. So `SinonSpy<[string, number], boolean>` is not assignable to `SinonSpy<any[], any>` and you couldn't pass it here if we didn't tell TypeScript to accept _any_ parameter type here. Note this doesn't degrade type safety in any way, this is just the parameter constraint, which is accurate.

Reply:
its not quite right IMO. `TArgs` is meant to be array-like to satisfy the fact that the `args` property is in fact an array on spies. if you do this, you're saying `args: any` which isn't quite right as it most definitely is always an array in sinon.

Reply:
`SinonSpy` still has the constraint `SinonSpy<TArgs extends any[]>`. So `spy.args` is still _always_ an array, unless you explicitly do this:
```ts
const spy: SinonSpy<any> = sinon.spy()
```
which you could already do before. Nothing changes except that you can now pass a `SinonSpy` with tuple `TArgs` to assertion functions which was a type error before (and it shouldn't be). You still can't pass a `SinonSpy` with non-array `TArgs` because it is still impossible to construct such a type due to the constraint `TArgs extends any[]`. You could only do that by casting to an instance of `SinonSpy<any>` as above, which you could already do before (because `any` is always assignable to anything, even if the constraint says `spy: SinonSpy<any[]>`), and this is expected if you explicitly use `any`.

This is the only way to achieve this with strictGenericChecks. TypeScript does not have the option to indicate whether a generic parameter should be covariant or contravariant like C# does.

I'm open to change my mind if you give me an example where this degrades type safety compared to before (I don't see one).

Reply:
`any` pretty much just disables the type checker, which is why it will happily satisfy `extends any[]` here. the fact that you've had to do `SinonSpy<any>` at all is the questionable part, i'm sure there's a better solution to allow tuples.

you're getting around the fact that a tuple isn't really an array type by turning off the type checker, effectively.

when i have time im happy to have a think about a better solution. because even though these are parameters and _technically_ dont change "anything", its still a hack in that you've disabled the type checker to get around a problem you had.

Reply:
would you mind reproducing the problem like [here](https://www.typescriptlang.org/play/index.html#src=interface%20Test%3CTArgs%20extends%20any%5B%5D%20%3D%20any%5B%5D%3E%20%7B%0D%0A%20%20args%3A%20TArgs%3B%0D%0A%7D%0D%0A%0D%0Aconst%20foo%3A%20Test%3C%5Bnumber%2C%20string%5D%3E%20%3D%20%7B%0D%0A%20%20args%3A%20%5B1%2C%20'2'%5D%0D%0A%7D%3B%0D%0A%0D%0Afunction%20thing(p%3A%20Test)%3A%20void%20%7B%20%7D%0D%0A%0D%0Athing(foo)%3B) as this example seems to type-check.

Reply:
> would you mind reproducing the problem like here as this example seems to type-check.

You are missing the call signature of `SinonSpy`. `args` is not important, important is the call signature, because that is checked contravariantly (assuming `strictFunctionTypes` is enabled). [Example](https://www.typescriptlang.org/play/index.html#src=interface%20Test%3CTArgs%20extends%20any%5B%5D%20%3D%20any%5B%5D%3E%20%7B%0D%0A%20%20(...args%3A%20TArgs)%3A%20any%0D%0A%7D%0D%0A%0D%0Aconst%20foo%3A%20Test%3C%5Bnumber%2C%20string%5D%3E%20%3D%20(n%3A%20number%2C%20s%3A%20string)%20%3D%3E%20%7B%20%7D%3B%0D%0A%0D%0Afunction%20thing(p%3A%20Test)%3A%20void%20%7B%20%7D%0D%0A%0D%0Athing(foo)%3B) (make sure to enable strictFunctionTypes in options).

The only alternative would be to splitting everything but the call signature into a separate interface `SinonSpyInspectable` (since `sinon.assert` does not need to call the spy, just inspect it), then accepting that interface for all the assertion functions. Do you think that is worth it?

Reply:
`notCalled<TArgs extends any[]>(spy: SinonSpy<TArgs>)`?

Reply:
That seems to work

Reply:
Hmm, it doesn't work for assertion functions with rest parameters though:

```ts
        callOrder<TArgs extends any[]>(...spies: Array<SinonSpy<TArgs>>): void;
```
we can't enforce that all spies have the same `TArgs`, and TS can't infer different types because there is nothing like rest type parameters.

Reply:
yup im struggling to think of a solution to that one, generic rest parameters...

we should keep the others as its nicer than forcing `any` but this one im not too sure about